lua/cmp/view/custom_entries_view.lua

```diff
@@ -172,12 +174,25 @@ custom_entries_view.open = function(self, offset, entries)
   local border_info = window.get_border_info({ style = completion })
   local border_offset_row = border_info.top + border_info.bottom
   local border_offset_col = border_info.left + border_info.right
-  if math.floor(vim.o.lines * 0.5) <= row + border_offset_row and vim.o.lines - row - border_offset_row <= math.min(DEFAULT_HEIGHT, height) then
+
+  local prefers_above = c.view.entries.vertical_positioning == 'above'
+  local prefers_auto = c.view.entries.vertical_positioning == 'auto'
+  local cant_fit_at_bottom = vim.o.lines - row - border_offset_row <= math.min(DEFAULT_HEIGHT, height)
+  local cant_fit_at_top = row - border_offset_row <= math.min(DEFAULT_HEIGHT, height)
+  local is_in_top_half = math.floor(vim.o.lines * 0.5) > row + border_offset_row
+  local should_position_above =
+    cant_fit_at_bottom or
+    (prefers_above and not cant_fit_at_top) or
+    (prefers_auto and is_in_top_half)
+  if should_position_above then
+    self.bottom_up = true
     height = math.min(height, row - 1)
     row = row - height - border_offset_row - 1
     if row < 0 then
       height = height + row
     end
+  else
+    self.bottom_up = false
   end
```

Comment:
Would be cool if this code some how took into account the experimental ghost text feature. Not sure if it can be done, but i would like to see auto checking if the ghost text is more than a number of lines that could maybe be configured as well. Could also just be set to 1 line.

Reply:
I guess you mean #1955

Reply:
Ye that about covers it :)